File: ios/Classes/AwareManager.swift
//
//  AwareManager.swift
//  whiteboard
//
//  Created by vicky Leu on 2021/5/19.
//

import Foundation
import Masonry
import WebKit

public class AwareManager: NSObject, BoardAwareInterface {
    let mTicManager: TICManager = TICManager.sharedInstance()
    var boardAware: BoardAware?
    var nativeViewLink: NativeViewLink?
    private let settingCallback = MySettingCallback()
    public var flutterApi: FLTPigeonFlutterApi?
    var drawerType = DrawerType.drawGraffiti
    var isHaveBackgroundImage=false
    public override init() {
        super.init()
        settingCallback.also { (it: MySettingCallback) in
            it.awareManager = self
        }
    }
    


    func preJoinClassroom(arg: FLTPreJoinClassRequest, ticCallback: @escaping TICCallback) {
        boardAware?.destroy()
        boardAware = BoardAware()
        mTicManager.`init`(Int32(arg.appId!.intValue), userId: arg.userId, userSig: arg.userSig)
        mTicManager.sendCommandBlock = { (ext, d) in
            let data = FLTReceivedData()
            data.data = FlutterStandardTypedData(bytes: d!)
            data.extension = ext
            print("receive嘿嘿嘿嘿嘿嘿嘿嘿呵呵呵呵呵呵::::\(data.data)")
            self.flutterApi?.receive(data, completion: { model, _ in
                if (model.code?.intValue == -1) {
                    print("同步失败了:${it.msg}")
                } else {
                    guard  let wtf: TEduBoardController = self.boardAware?.mBoard else {
                        return
                    }
                    //                    wtf.addAckData(data) ////
                }
            })
        }
        //1、设置白板的回调
        let mbcallback = MyBoardCallback(self)
        boardAware?.mBoardCallback = mbcallback
        
        ticCallback(TICModule.TICMODULE_IMSDK, 1, "预创建参数初始化成功")
    }
    
    func joinClass(_ classroomOption: TICClassroomOption, ticCallback: @escaping TICCallback) {
        classroomOption.boardDelegate = boardAware?.mBoardCallback
        mTicManager.initTEduBoard(classroomOption)
        guard let boardController = mTicManager.getBoardController() else {
            ticCallback(TICModule.TICMODULE_IMSDK, -1, "白板初始化不成功")
            return
        }
        boardController.setGlobalBackgroundColor(UIColor.clear)
        boardAware?.mBoard = boardController
        classroomOption.boardDelegate?.onTEBInit()
        
        
        ///腾讯的Android和iOS回调不同步,这里手动调用,保证在业务层处理逻辑是一样的.反正回调中我会判断画板是否已经准备就绪的
        ticCallback(TICModule.TICMODULE_IMSDK, 1, "创建课堂 成功, 房间号 \(classroomOption.classId)")
    }
    
    //////
    
    func drawGraffiti() {
        guard (drawerType != DrawerType.drawGraffiti) else {
            return
        }
        drawerType = DrawerType.drawGraffiti
        boardAware?.mBoard?.setToolType(TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_PEN)
    }
    
    func drawLine() {
        guard (drawerType != DrawerType.drawLine) else {
            return
        }
        drawerType = DrawerType.drawLine
        boardAware?.mBoard?.setToolType(TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_LINE)
    }
    
    func drawSquare() {
        guard (drawerType != DrawerType.drawSquare) else {
            return
        }
        drawerType = DrawerType.drawSquare
        boardAware?.mBoard?.setToolType(TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_RECT)
    }
    
    func drawCircular() {
        guard (drawerType != DrawerType.drawCircular) else {
            return
        }
        drawerType = DrawerType.drawCircular
        boardAware?.mBoard?.setToolType(TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_OVAL)
    }
    
    func drawText() {
        guard (drawerType != DrawerType.drawText) else {
            return
        }
        drawerType = DrawerType.drawText
        boardAware?.mBoard?.setTextStyle(TEduBoardTextStyle.TEDU_BOARD_TEXT_STYLE_NORMAL)
        boardAware?.mBoard?.setToolType(TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_TEXT)
    }
    
    func eraserDrawer() {
        guard (drawerType != DrawerType.eraserDrawer) else {
            return
        }
         drawerType = DrawerType.eraserDrawer
        let arr : Array<NSNumber> = [
            NSNumber(value:TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_LINE.rawValue),
            NSNumber(value:TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_OVAL.rawValue),
            NSNumber(value:TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_MOUSE.rawValue),
            NSNumber(value:TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_PEN.rawValue),
            NSNumber(value:TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_RECT.rawValue),
            NSNumber(value:TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_TEXT.rawValue),
        ]
        boardAware?.mBoard?.setToolType(TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_ERASER)
        boardAware?.mBoard?.setEraseLayerType(arr)
    }
    
    func rollbackDraw() {
        boardAware?.mBoard?.undo()
    }
    
    func wipeDraw() {
        boardAware?.mBoard?.clearDraws()
        isHaveBackgroundImage = false
    }

    func removeBackgroundImage(){
        boardAware?.mBoard?.clearBackground(true, andSelected: false)
        isHaveBackgroundImage = false
    }


    func setToolColor(color:UIColor){
        switch drawerType {
        case .drawGraffiti:
            boardAware?.mBoard?.setBrush(color)
            break
        case .drawLine:
            boardAware?.mBoard?.setBrush(color)
            break
        case .drawSquare:
            boardAware?.mBoard?.setBrush(color)
            break
        case .drawCircular:
            boardAware?.mBoard?.setBrush(color)
            break
        case .drawText:
            boardAware?.mBoard?.setTextColor(color)
            break
        case .eraserDrawer:
            break
        }
    }
    func setToolSize(size:Int){
        switch drawerType {
        case .drawGraffiti:
            boardAware?.mBoard?.setBrushThin(UInt32(size))
            break
        case .drawLine:
            boardAware?.mBoard?.setBrushThin(UInt32(size))
            break
        case .drawSquare:
            boardAware?.mBoard?.setBrushThin(UInt32(size))
            break
        case .drawCircular:
            boardAware?.mBoard?.setBrushThin(UInt32(size))
            break
        case .drawText:
            boardAware?.mBoard?.setTextSize(UInt32(size))
            break
        case .eraserDrawer:
            break
        }
    }
    
    //////
    
    func onTEBSyncData(data: String) {
        let model = FLTReceivedData()
        model.data = FlutterStandardTypedData(bytes: data.data(using: .utf8)!)
        model.extension = "TXWhiteBoardExt"
        print("receive嘿嘿嘿嘿嘿嘿嘿嘿呵呵呵呵呵呵:::\(model.data)")
        self.flutterApi?.receive(model, completion: { model, _ in
            if (model.code?.intValue == -1) {
                print("同步失败了:${it.msg}")
            }
        })
    }
    
    func reset() {
        boardAware?.reset()
        drawerType = DrawerType.drawGraffiti
        isHaveBackgroundImage = false
    }
    
    func setBackgroundColor(_ color: UIColor) {
        boardAware?.setBackgroundColor(color)
    }
    
    func addBackgroundImage(url: String) {
        boardAware?.mBoard?.setBackgroundImage(url, mode: TEduBoardImageFitMode.TEDU_BOARD_IMAGE_FIT_MODE_CENTER)
        isHaveBackgroundImage = true
    }
    
    func quitClassroom() {
        boardAware?.destroy()
        isHaveBackgroundImage = false
        mTicManager.quitClassroom(true, callback: { _, errCode, errMsg in
            if (errCode == -1) {
                print("mother fucker  退出白板失败:\(errCode)  \(errMsg)")
                self.removeBoardView()
            } else {
                self.removeBoardView()
            }
        })
        mTicManager.sendCommandBlock = nil
        flutterApi?.exitRoom(FLTDataModel().also { (v: FLTDataModel) in
            v.code = 1
            v.msg = "退出成功"
            v.data = nil
        }) { (_, error: Error?) in
            
        }
    }
    
    
    func receiveData(data: [UInt8], callback: @escaping TICCallback) {
        do {
            let string = try String(bytes: data, encoding: .utf8)
            try boardAware?.mBoard?.addSyncData(string)
            callback(TICModule.TICMODULE_IMSDK, 1, "同步数据成功")
        } catch let error {
            callback(TICModule.TICMODULE_IMSDK, -1, "addSyncData failed: \(error)")
        }
    }
    
    
    /////
    func onTEBHistroyDataSyncCompleted() {
        flutterApi?.historySyncCompleted { _, _ in
            
        }
        guard let board = boardAware?.mBoard else {
            return
        }
        guard let currentBoard = board.getCurrentBoard() else {
            return
        }
        let currentFile = board.getCurrentFile()
        print("DataSyncCompleted currentBoard:\(currentBoard) currentFile:\(currentFile)")
    }
    
    func addBoardView() {
        guard let board = boardAware?.mBoard else {
            return
        }
        guard let boardView: WKWebView = board.getBoardRenderView() as? WKWebView else {
            return
        }
        print("addBoardView::::::\(boardView)")
        if #available(iOS 11.0, *) {
            boardView.scrollView.contentInsetAdjustmentBehavior = .never
        }
        boardView.scrollView.contentInset = .zero
        boardView.scrollView.backgroundColor = UIColor.clear
        boardView.isOpaque = false
        boardView.backgroundColor = UIColor.clear
        
        nativeViewLink?.addView(boardView) { (root: UIView, make: MASConstraintMaker?) in
            make?.top.equalTo()(root)
            make?.left.equalTo()(root)
            make?.right.equalTo()(root)
            make?.bottom.equalTo()(root)
        }
    }
    
    func removeBoardView() {
        guard let board = boardAware?.mBoard else {
                return
            }
        guard let boardView = board.getBoardRenderView() else {
            return
        }
        nativeViewLink?.removeView(boardView)
        board.unInit()
    }
    
    func setCanUndo(_ canUndo: Bool) {
        settingCallback.setCanUndo(canUndo)
    }
    
    func setCanRedo(_ canredo: Bool) {
        settingCallback.setCanRedo(canredo)
    }
    
    func addFile(_ fileId: String?) -> TEduBoardFileInfo? {
        return nil
    }
    
    func onTextComponentStatusChange(_ id: String?, _ status: String?) {
        //
    }
    
    func receiveIds(id: String, type: Int) {
        //rtcAware?.mImgsFid
    }
}

enum DrawerType {
    case drawGraffiti
    case drawLine
    case drawSquare
    case drawCircular
    case drawText
    case eraserDrawer
}

File: android/src/main/kotlin/com/bond/whiteboard/AwareManager.kt
package com.bond.whiteboard

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Build
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout
import androidx.annotation.ColorInt
import com.bond.whiteboard.board.BoardAware
import com.bond.whiteboard.nativeView.NativeViewLink
import com.bond.whiteboard.teb.BoardAwareInterface
import com.bond.whiteboard.teb.MyBoardCallback
import com.pigeon.PigeonPlatformMessage
import com.tencent.smtt.sdk.WebView
import com.tencent.teduboard.TEduBoardController
import com.tencent.teduboard.TEduBoardController.TEduBoardFileInfo
import com.tencent.tic.core.TICCallback
import com.tencent.tic.core.TICClassroomOption
import com.tencent.tic.core.TICManager
import com.tencent.tic.core.TICManager.Companion.MODULE_IMSDK
import com.tencent.tic.core.TICManager.Companion.TICSDK_WHITEBOARD_CMD
import com.tencent.tic.core.TICManager.TICIMStatusListener


class AwareManager : TICIMStatusListener, BoardAwareInterface {
    var nativeViewLink: NativeViewLink?=null
    var flutterApi: PigeonPlatformMessage.PigeonFlutterApi?=null
    var drawerType = DrawerType.drawGraffiti
    var isHaveImageBackground = false
    private val settingCallback = MySettingCallback().also {
        it.awareManager=this
    }
    /**
     * 课堂资源互动管理
     */
    protected var mTicManager: TICManager = TICManager.instance.also {
        it.receiveData(this)
    }

    var boardAware: BoardAware? = null
    fun preJoinClassroom(
        arg: PigeonPlatformMessage.PreJoinClassRequest,
        ticCallback: TICCallback<Any>
    ) {
        val context = nativeViewLink?.getApplicationContext()
        if(context==null){
            val msg="预创建房间失败,上下文为空"
            print(msg)
            ticCallback.onError(MODULE_IMSDK,-1,msg)
            return
        }
        boardAware?.destroy()
        boardAware=BoardAware(context)
        try {
            mTicManager.init(context,arg.appId.toInt(), arg.userId, arg.userSig)
        }catch (e:Exception){
            Log.e("mother fucker","why you are crash anytime")
        }
        //2.白板
        val board = mTicManager.boardController
        if(board==null){
            val msg="预创建房间失败,白板未初始化"
            print(msg)
            ticCallback.onError(MODULE_IMSDK,-1,msg)
            return
        }
        board.globalBackgroundColor= TEduBoardController.TEduBoardColor(Color.TRANSPARENT)
        boardAware?.mBoard = board
        //1、设置白板的回调
        boardAware?.mBoardCallback = MyBoardCallback(this)
        print("预创建参数初始化成功")
        ticCallback.onSuccess(1)
    }
    fun joinClassroom(
        classroomOption: TICClassroomOption,
        ticCallback: TICCallback<Any>
    ) {
        classroomOption.boardCallback = boardAware?.mBoardCallback
        mTicManager.initTEduBoard(classroomOption)
        print("创建课堂 成功, 房间号：${classroomOption.classId}")
        ticCallback.onSuccess(1)
    }

    fun reset() {
        boardAware?.reset()
        drawerType = DrawerType.drawGraffiti
        isHaveImageBackground=false
    }
    fun setBackgroundColor(@ColorInt color:Int) {
        boardAware?.setBackgroundColor(color)
    }
    fun quitClassroom() {
        boardAware?.destroy()
        isHaveImageBackground=false
        mTicManager.quitClassroom(true,object :TICCallback<Any>{
            override fun onSuccess(data: Any) {
                removeBoardView()
            }
            override fun onError(module: String, errCode: Int, errMsg: String) {
                Log.e("mother fucker","退出白板失败:$errCode  $errMsg")
                removeBoardView()
            }
        })
//        rtcAware?.destroy()
        flutterApi?.exitRoom(PigeonPlatformMessage.DataModel().apply {
            this.code=1
            this.msg="退出成功"
            this.data=null
        }) {

        }
    }


    fun receiveData(data: ByteArray, callback: TICCallback<Any>) {
        try {
            boardAware?.mBoard?.addSyncData(String(data))
            callback.onSuccess(1)
        }catch (e:Exception){
            callback.onError(TICManager.MODULE_IMSDK, -1, "addSyncData failed: ${e.message}")
        }

    }

    override fun onTICForceOffline() {

    }

    override fun onTICUserSigExpired() {
        //
    }
    override fun onTEBHistroyDataSyncCompleted() {
        flutterApi?.historySyncCompleted(null)
        val board = boardAware?.mBoard ?: return
        val currentBoard: String = board.currentBoard;
        val currentFile = board.currentFile
        Log.w("DataSyncCompleted", "currentBoard: $currentBoard currentFile: $currentFile")
    }

    override fun addBoardView() {
        val board= boardAware?.mBoard ?:return
        board.backgroundColor= TEduBoardController.TEduBoardColor(Color.TRANSPARENT)
        val boardView= board.boardRenderView ?:return
        val webView = boardView as WebView
        if(Build.VERSION.SDK_INT>=Build.VERSION_CODES.LOLLIPOP) {
            try {
                webView.settings.mixedContentMode=0
            }catch (e:Exception){}
        }
        webView.isFocusable=true
        webView.isFocusableInTouchMode=true

        webView.requestFocus(View.FOCUS_DOWN)
        webView.setOnTouchListener { v,  event ->
            when (event.getAction()) {
                MotionEvent.ACTION_DOWN, MotionEvent.ACTION_UP -> if (!v.hasFocus()) {
                    v.requestFocus()
                }
            }
            false
        }

        webView.setOnFocusChangeListener { v, hasFocus ->
            Log.e("setOnFocusChangeLis","${hasFocus}")
            v.requestFocus()
        }
        webView.setBackgroundColor(Color.TRANSPARENT)
        webView.setPadding(0,0,0,0)
        val layoutParams = FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT)
        nativeViewLink?.addView(boardView,layoutParams)
    }
    override fun removeBoardView() {
        val board= boardAware?.mBoard  ?:return
        val boardView= board.boardRenderView ?:return
        nativeViewLink?.removeView(boardView)
        board.uninit()
    }
    override fun setCanUndo(canUndo: Boolean) {
        settingCallback.setCanUndo(canUndo)
    }
    override fun setCanRedo(canredo: Boolean) {
        settingCallback.setCanRedo(canredo)
    }
    override fun addFile(fileId: String?): TEduBoardFileInfo? {
        return null
    }


    fun addBackgroundImage(url: String) {
        boardAware?.mBoard?.setBackgroundImage(url, TEduBoardController.TEduBoardImageFitMode.TEDU_BOARD_IMAGE_FIT_MODE_CENTER)
        isHaveImageBackground=true
    }


    override fun onTextComponentStatusChange(id: String?, status: String?) {
        //
    }


    override fun sendMessage(data: ByteArray, extension: String) {
        flutterApi?.receiveData(PigeonPlatformMessage.ReceivedData().also {
            it.data=data
            it.extension=extension
        }){it->
            if(it.code.toInt()==-1){
                print("同步失败了:${it.msg}")
            }else{
                val wtf : TEduBoardController = boardAware?.mBoard?:return@receiveData
//                wtf.addAckData(data) //// 🙄🙄🙄🙄🙄🙄🙄❓❓❓❓❓❓❓❓❓❓🙃🙃🙃🙃🙃🙃🙃🙃🙃🙃🙃
            }
        }
    }
    override fun onTEBSyncData(data: String) {
        sendMessage(data.toByteArray(),extension = TICSDK_WHITEBOARD_CMD)
    }

    fun receiveIds(id: String, type: Int) {
//        rtcAware?.mImgsFid
    }

    fun drawGraffiti() {
        if(drawerType==DrawerType.drawGraffiti)return
        drawerType=DrawerType.drawGraffiti
        boardAware?.mBoard?.toolType = TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_PEN
    }

    fun drawLine() {
        if(drawerType==DrawerType.drawLine)return
        drawerType=DrawerType.drawLine
        boardAware?.mBoard?.toolType = TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_LINE
    }

    fun drawSquare() {
        if(drawerType==DrawerType.drawSquare)return
        drawerType=DrawerType.drawSquare
        boardAware?.mBoard?.toolType = TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_RECT
    }

    fun drawCircular() {
        if(drawerType==DrawerType.drawCircular)return
        drawerType=DrawerType.drawCircular
        boardAware?.mBoard?.toolType = TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_OVAL
    }

    fun drawText() {
        if(drawerType==DrawerType.drawText)return
        drawerType=DrawerType.drawText
        boardAware?.mBoard?.textStyle = TEduBoardController.TEduBoardTextStyle.TEDU_BOARD_TEXT_STYLE_NORMAL
        boardAware?.mBoard?.toolType = TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_TEXT
    }

    fun eraserDrawer() {
        if(drawerType==DrawerType.eraserDrawer)return
        drawerType=DrawerType.eraserDrawer
        val arr : ArrayList<Int> = arrayListOf<Int>(
            TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_LINE,
            TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_OVAL,
            TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_MOUSE,
            TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_PEN,
            TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_RECT,
            TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_TEXT)
        boardAware?.mBoard?.toolType = TEduBoardController.TEduBoardToolType.TEDU_BOARD_TOOL_TYPE_ERASER
        boardAware?.mBoard?.setEraseLayerType(arr)
    }

    fun rollbackDraw() {
        boardAware?.mBoard?.undo()
    }

    fun wipeDraw() {
        boardAware?.mBoard?.clear(false)
        isHaveImageBackground=false
    }

    fun setToolColor(value: String) {
        val color=TEduBoardController.TEduBoardColor(value)
        when(drawerType) {
            DrawerType.drawGraffiti->{
                boardAware?.mBoard?.brushColor=color
            }
            DrawerType.drawLine->{
                boardAware?.mBoard?.brushColor=color
            }
            DrawerType .drawSquare->{
                boardAware?.mBoard?.brushColor=color
            }
            DrawerType .drawCircular->{
                boardAware?.mBoard?.brushColor=color
            }
            DrawerType .drawText->{
                boardAware?.mBoard?.textColor=color
            }
            DrawerType .eraserDrawer->{
            }
        }
    }

    fun setToolSize(size: Int) {
        when(drawerType) {
            DrawerType.drawGraffiti->{
                boardAware?.mBoard?.brushThin = size
            }
            DrawerType .drawLine->{
                boardAware?.mBoard?.brushThin = size
            }
            DrawerType .drawSquare->{
                boardAware?.mBoard?.brushThin = size
            }
            DrawerType .drawCircular->{
                boardAware?.mBoard?.brushThin = size
            }
            DrawerType .drawText->{
                boardAware?.mBoard?.textSize = size
            }
            DrawerType .eraserDrawer->{
            }
        }
    }


    fun removeImageBackground() {
        boardAware?.mBoard?.clear(true)
        isHaveImageBackground=false
    }

}

enum class DrawerType{
    drawGraffiti,
    drawLine,
    drawSquare,
    drawCircular,
    drawText,
    eraserDrawer,
}